src/pcd_db_dets.erl

```diff
@@ -0,0 +1,347 @@
+%%  case get_riak_pid(Owner) of
+%%    Pid when is_pid(Pid) ->
+%%      BinKey = encode_key(Key, Converter),
+%%      riakc_pb_socket:fetch_type(Pid, Bucket, BinKey);
+%%    WAFIT ->
+%%      _ = lager:error("Riak connection. Reason:~p", [WAFIT]),
+%%      {error, WAFIT}
+%%  end.
+%%
+%%update_type(Bucket, Key, Operation, Owner, Converter) ->
+%%  update_type(Bucket, Key, Operation, Owner, Converter, []).
+%%
+%%update_type(Bucket, Key, Operation, Owner, Converter, Options) ->
+%%  case get_riak_pid(Owner) of
+%%    Pid when is_pid(Pid) ->
+%%      BinKey = encode_key(Key, Converter),
+%%      riakc_pb_socket:update_type(Pid, Bucket, BinKey, Operation, Options);
+%%    WAFIT ->
+%%      _ = lager:error("Riak connection cannot be set up. Reason:~p", [WAFIT]),
+%%      {error, WAFIT}
+%%  end.
+
+close() ->
+  ok.
+
+%% ====================================================================
+%% Internal functions
+%% ====================================================================
+
+encode_key(Key, none) -> term_to_binary(Key);
+encode_key(Key, _Converter) when is_binary(Key) -> Key;
+encode_key(Key, Converter) ->
+  {ok, ConvertedKey} = Converter:to_json(Key),
+  ConvertedKey.
+
+%% encode_value/1
+%% <a>Encode value with encoder if possible</a>
+-spec encode_value(Value:: term(), Converter :: atom()) -> Result when
+  Result :: {ContentType :: string(), Value :: term()}.
+
+encode_value(Value, _Converter) when is_binary(Value) ->
+  {"application/x-erlang-binary", term_to_binary(Value)};
+encode_value(Value, Converter) ->
+  case io_lib:printable_unicode_list(Value) of
+    true ->
+      {"application/text", list_to_binary(Value)};
+    _ ->
+      case is_tuple(Value) of
+        true ->
+          encode_from_tuple(Value, Converter);
+        _ ->
+          % not tuple just convert it to binary
+          {"application/x-erlang-binary", term_to_binary(Value)}
+      end
+  end.
+
+encode_from_tuple(Value, Converter) ->
+  try element(1, Value) of
+    RecName when is_atom(RecName) ->
+      case is_record(Value, RecName) of
+        true ->
+          try_to_convert_to_json(Value, Converter);
+        _ ->
+          % record is not in a format to convert
+          {"application/x-erlang-binary", term_to_binary(Value)}
+      end;
+    _ ->
+      _ = lager:warning(
+        "Record format is not suitable to convert. Value:~p. " ++
+          "Converting term to binary",
+        [Value]),
+      {"application/x-erlang-binary", term_to_binary(Value)}
+  catch
+    A:B ->
+      _ = lager:error("Exception:~p:~p"
+                      "Record format is not suitable"
+                      " to convert. Record:~p. "
+                      "Stack:~p", [A, B, Value,
+                                   erlang:get_stacktrace()]),
+      {"application/x-erlang-binary", term_to_binary(Value)}
+  end.
+
+
+try_to_convert_to_json(Value, Converter) ->
+  try
+    {ok, EncodedValue} = Converter:to_json(Value),
```

Comment:
Remove the dynamic function call on line 308. Only modules that define callbacks should make dynamic calls.